In [1]:
%pip install plotly -q

Note: you may need to restart the kernel to use updated packages.


In [2]:
import plotly.express as px
import plotly.graph_objects as go
import numpy as np
from sklearn.preprocessing import StandardScaler

# Base idade e salário

In [3]:
from sklearn.cluster import KMeans

In [4]:
x = [20,  27,  21,  37,  46, 53, 55,  47,  52,  32,  39,  41,  39,  48,  48]
y = [1000, 1200, 2900, 1850, 900, 950, 2000, 2100,
     3000, 5900, 4100, 5100, 7000, 5000, 6500]

In [5]:
graph = px.scatter(x = x, y = y)
graph.show()

In [7]:
base_salary = np.array([
  [age, salary]
  for age, salary in zip(x, y)
])
base_salary

array([[  20, 1000],
       [  27, 1200],
       [  21, 2900],
       [  37, 1850],
       [  46,  900],
       [  53,  950],
       [  55, 2000],
       [  47, 2100],
       [  52, 3000],
       [  32, 5900],
       [  39, 4100],
       [  41, 5100],
       [  39, 7000],
       [  48, 5000],
       [  48, 6500]])

In [8]:
scaler_salary = StandardScaler()
base_salary = scaler_salary.fit_transform(base_salary)

In [9]:
base_salary

array([[-1.87963884, -1.11413572],
       [-1.23255006, -1.01725435],
       [-1.78719758, -0.19376273],
       [-0.30813751, -0.70238991],
       [ 0.52383377, -1.1625764 ],
       [ 1.17092255, -1.13835606],
       [ 1.35580506, -0.62972888],
       [ 0.61627503, -0.5812882 ],
       [ 1.0784813 , -0.14532205],
       [-0.77034379,  1.25945777],
       [-0.12325501,  0.38752547],
       [ 0.0616275 ,  0.8719323 ],
       [-0.12325501,  1.79230528],
       [ 0.70871628,  0.82349162],
       [ 0.70871628,  1.55010187]])

In [10]:
kmeans_salary = KMeans(n_clusters=3)
kmeans_salary.fit(base_salary)

/home/gabriel/Projects/machine-learning-e-data-science-com-python/.venv/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:1412: FutureWarning:

The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning



KMeans(n_clusters=3)

In [12]:
centroides = kmeans_salary.cluster_centers_
centroides

array([[ 0.07703438,  1.11413572],
       [ 0.73953003, -0.72661025],
       [-1.63312883, -0.77505093]])

In [13]:
scaler_salary.inverse_transform(kmeans_salary.cluster_centers_)

array([[  41.16666667, 5600.        ],
       [  48.33333333, 1800.        ],
       [  22.66666667, 1700.        ]])

In [14]:
labels = kmeans_salary.labels_
labels

array([2, 2, 2, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0], dtype=int32)

Explicação dos rótulos (labels):

- 2 representa o grupo com idade de $22.66666667$ e salário de $1700$
- 1 representa o grupo com idade de $48.33333333$ e salário de $1800$
- 0 representa o grupo com idade de $41.16666667$ e salário de $5600$

In [16]:
graph1 = px.scatter(x = base_salary[:, 0], y = base_salary[:, 1], color=labels)
graph2 = px.scatter(x = centroides[:, 0], y = centroides[:, 1], size=[12, 12, 12])
graph3 = go.Figure(data = graph1.data + graph2.data)
graph3.show()